### PR TITLE
Enable Plugins to create custom routes and middlewares

### PR DESCRIPTION
**Pull request by RoeHH**

### CODE REVIEW

**Comment by RoeHH** [comment]
Hey a review would be appreciated. You can read what i added in the Plugin.md
@lucacasonato :)

File: src/server/types.ts [hunk 1]
```diff
@@ -304,3 +308,20 @@ export interface PluginRenderFunctionResult {
    */
   requiresHydration: boolean;
 }
+
+export interface PluginMiddlewear {
```
Comment (thegaryroberts) [inline src/server/types.ts]: typo: Middlewear = Middleware

**Comment by vibegui** [comment]
We need this badly at deco.cx!! Can we help?

File: plugins/routplug.tsx [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+function Home(props: PageProps) {
+  const { test } = props.params;
+
+  console.log("hi handler", test);
+  return (
+    <div>
+      <p>
+        Welcome to Fresh. Try to update this message in the ./routes/index.tsx
+        file, and refresh.666666
```
Comment (thegaryroberts) [inline plugins/routplug.tsx]: ? https://www.youtube.com/watch?v=L25NgHpsMsA

Comment (RoeHH) [reply plugins/routplug.tsx]: ohh i should remove this file i just used it to test the fature localy

**Comment by RoeHH** [comment]
@guifromrio Sorry didn't see it, I think we would need help from someone with write access to resolve the merge conflicts

**Comment by bastilian** [comment]
@RoeHH You should be able to resolve this conflict as well. If you add this, the main repo, as a remote to your local working copy and rebase this branch on top of the latest changes it will show you the conflicts and an opportunity to fix them. Currently it seems that the conflict is with the `.gitignore` file, which has been remove in `main`. If you remove the ignore file from this branch too, the conflict should be gone as well. However, rebasing would probably be good either way.

**Comment by marvinhagemeister** [comment]
Closing as the changes here were included in #1455 🎉

**Pull request closed without merging**